### ISSUE TITLE
Insert mode doesn't work properly in duckduckgo.com when input.insert_mode.auto_enter = False

### ISSUE THREAD

**Issue by prosoitos** [qutebrowser/qutebrowser#4600]
**Version info**

qutebrowser v1.5.2
Git commit: [b3e1bac](https://github.com/qutebrowser/qutebrowser/commit/b3e1bac11b08b0a25d227123c9fb21b60c6970cd) (2019-02-16 23:19:28 +0100)
Backend: QtWebEngine (Chromium 69.0.3497.128)

CPython: 3.7.2
Qt: 5.12.1
PyQt: 5.12

**Does the bug happen if you start with `--temp-basedir`?**

No

**Description**

In the duckduckgo.com search box (and only on that site, and only after an initial search--so when the search box is at the top of the screen), `insert-mode` gets dysfunctional: as I type, instead of inserting characters, the keys do what they would do if I were not in insert mode. This happens while the `-- INSERT MODE --` bar is still present. Typing extremely slowly helps, but it still jumps.

I mentioned it a while ago in #2562, but I think that it had nothing to do whatsoever with the issue in question and was simply silly of me to write about this there.

I think that the problem is caused by `c.input.insert_mode.auto_enter = False`: I haven't tested it extensively, but I don't seem to have the problem with `qutebrowser --temp-basedir`, nor by reverting `input.insert_mode.auto_enter` to default.

I still prefer to have `c.input.insert_mode.auto_enter = False` though because `insert-mode` is not entered automatically as it should on some websites (e.g. Facebook) and the unpredictability was making me make the mistake of typing without being in `insert-mode` all the times. Having to enter `insert-mode` manually, while not ideal, is at least consistant and prevents me from making that mistake.

The duckduckgo problem has been going on for some time (2 months maybe????) and is pretty bad: if i want to modify a search by adding or editing anything, I have to retype the whole thing with `:open`. Even typing a few characters extremely slowly on the search box is sketchy.

**Comment by user202729** [comment]
Can you post some debug log or your configuration file?

**Comment by jgkamat** [comment]
Marie-Helene Burle writes:

> In the duckduckgo.com search box (and only on that site, and only after an initial search--so when the search box is at the top of the screen), `insert-mode` gets dysfunctional: as I type, instead of inserting characters, the keys do what they would do if I were not in insert mode. This happens while the `-- INSERT MODE --` bar is still present. Typing extremely slowly helps, but it still jumps.

> I think that the problem is caused by `c.input.insert_mode.auto_enter = False`: I haven't tested it extensively, but I don't seem to have the problem with `qutebrowser --temp-basedir`, nor by reverting `input.insert_mode.auto_enter` to default.

I don't think that setting changes much, here's its only current usage:

https://github.com/qutebrowser/qutebrowser/blob/17846ea3862c2394679355d4d851b4b8bc813112/qutebrowser/browser/mouse.py#L186

can you reproduce with --temp-basedir and changing the auto_enter setting on
top of that?

I also cannot reproduce, FWIW.

**Comment by The-Compiler** [comment]
I can't reproduce either. What do you mean with "instead of inserting characters, the keys do what they would do if I were not in insert mode" precisely? Does e.g. typing `d` indeed close the current tab?

**Comment by prosoitos** [comment]
> What do you mean with "instead of inserting characters, the keys do what they would do if I were not in insert mode" precisely? Does e.g. typing `d` indeed close the current tab?

Yes, exactly. The `insert-mode` bar remains visible, but space will make me scroll down the page, etc. So typing becomes suddenly utter chaos.

**Comment by prosoitos** [comment]
> can you reproduce with --temp-basedir and changing the auto_enter setting on top of that?

I can't reproduce with that... so might be the auto_enter on top of other things in my settings...

**Comment by prosoitos** [comment]
> Can you post some debug log or your configuration file?

[config.txt](https://github.com/qutebrowser/qutebrowser/files/2888391/config.txt)

**Comment by prosoitos** [comment]
Note: to try to reproduce, make sure that you first do a search in duckduckgo, then edit the search by typing in the search box that will be at the top of the page.

My normal workflow is to search using `:open -t` (which works without any issue at all), then edit the search if need be. And it is that part that is just not working at all.

**Comment by The-Compiler** [comment]
Does the input box have keyboard focus (i.e., do you see a blinking cursor in it)? All qutebrowser does in insert mode is passing through keybindings to the page, so if the box doesn't have keyboard focus, I'd expect certain keys (Space, those handled by DuckDuckGo like cursor keys or hjkl, etc.) to work - but actual qutebrowser bindings (like, say, `;m` to open the messages with your config) won't.

**Comment by user202729** [comment]
`:log debug --raw` or something like that.

**Comment by prosoitos** [comment]
> I can't reproduce the issue even with the exact git commit and config file. Probably have something to do with your window manager.

Must be... #4579 seems to also be due to my WM (and to happen with another Lisp based WM)... :slightly_frowning_face:

**Comment by prosoitos** [comment]
> Does the input box have keyboard focus (i.e., do you see a blinking cursor in it)? All qutebrowser does in insert mode is passing through keybindings to the page, so if the box doesn't have keyboard focus, I'd expect certain keys (Space, those handled by DuckDuckGo like cursor keys or hjkl, etc.) to work - but actual qutebrowser bindings (like, say, `;m` to open the messages with your config) won't.

There is a blinking cursor in the box, yes.

I don't know for sure which keys create this or not as typing one single key at a time usually does not trigger it and when I type at a normal speed, all sorts of things happen at once. But Space triggering scrolling down is the one thing that I always notice. So I guess that you must be correct there (that only the keys handled by duckduckgo are problematic).

**Comment by prosoitos** [comment]
Oh, another one: the backspace stops working also. This one is much less disruptive since nothing crazy happens, but I can see that the cursor suddenly disappears and backspace has no effect (while insert-mode bar still visible).

Also, of note, when the scrolling happens with space, if I try to go back to the top of the page (`u` is my binding for that and I don't remember the default binding), it doesn't work (so the insert-mode is still active somehow).

**Comment by user202729** [comment]
I guess it's probably another kind of extra TAB. Just do `bind -m insert <tab> nop`

**Comment by prosoitos** [comment]
I had not understood your comment at all about the `<tab>` on #4579 because I had missed @codesections comment.

I added to my config file:

```py
c.bindings.commands = {
    'normal': {'<tab>': 'nop',},
    'insert': {'<tab>': 'nop',}
}
```
and it *seems* to have solved both problems. I will test a little longer and confirm.

**Comment by prosoitos** [comment]
Out of curiosity, I tried with only the first line and it is enough to solve the problem.

```python
c.bindings.commands = {
    'normal': {'<tab>': 'nop',}
}
```
Even if it is still very unclear what is going on, the workaround does the trip, so I am closing this for now.

**Comment by prosoitos** [comment]
A little bit of extra information, in case it might help @The-Compiler in figuring out those rogue TAB key presses:

Even if it doesn't make much sense, assigning `nop` to tab in *normal mode* had a huge effect in solving this problem (even though the problem is about insert mode). So much so that I thought that it was sufficient. However, though infrequently, the problem still occurs on occasion. Re-adding the line to assign `nop` to tab in insert mode definitely fixes the problem.

I thought I'd mention this observation, in case it may be useful (particularly since what I wrote above was not fully correct).

**Comment by prosoitos** [comment]
Ah! I think that I finally found what the problem was on duckduckgo!

On a duckduckgo page, going to "settings", then "other settings" and switching "Enables keyboard shortcuts on the site" to `OFF` seems to have finally stopped this entirely. I tested it without assigning TAB to `nop` and it works fine there too (I will put this option back as it also solves #4579, but has now become unnecessary for this issue.

Now, why am I the only one having the problem? It may be that `OFF` was the default option and that I changed it to `ON` when I started using duckduckgo (I can't remember, but I don't think so). Or it might be that the duckduckgo keybindings were interfering with the EXWM keybindings (more likely).

Either way, I think that this is the real fix for this problem.

**Comment by prosoitos** [comment]
The problem is still occurring (though not as bad as before binding `<tab>` to `nop`). Still an issue though.

On top of duckduckgo, I realized that it also happens on google maps.

**Comment by 20jam** [comment]
It descover the issue inside github search. ITS REALY BAD

**Comment by The-Compiler** [comment]
> It descover the issue inside github search. ITS REALY BAD

...what issue do you mean exactly?

**Comment by 20jam** [comment]
After I search for something jk/page movement stops working. Unless I press tab, which redirects the focus from search input to another page element, then jk page movement works. I was trying to ignore until I found my self having a painful experience with it. it also redirects focus to input when switching to next search result page . Auto focus issue I guess

**Comment by The-Compiler** [comment]
@20jam That sounds completely unrelated to this issue, and also I can't reproduce. Please open a new issue with the information requested there.

**Comment by 20jam** [comment]
Oh Sorry about that, Thank you.
